In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

df = pd.read_csv("../data/raw/customers.csv")

df.head()

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue
0,C000001,East,SMB,Social,3,2,13.4,0,5041.80
1,C000002,South,SMB,Partner,8,5,8.7,0,11299.93
2,C000003,West,Consumer,Organic,23,3,7.2,0,4540.92
3,C000004,East,Consumer,Organic,8,2,6.4,0,3303.00
4,C000005,North,Consumer,Referral,38,7,7.6,0,8129.65


In [2]:
df.shape

(10020, 9)

In [3]:
df.columns.tolist()

['customer_id',
 'region',
 'customer_type',
 'acquisition_channel',
 'tenure_months',
 'orders',
 'discount_pct',
 'returns',
 'revenue']

In [4]:
df.dtypes

customer_id                str
region                     str
customer_type              str
acquisition_channel        str
tenure_months            int64
orders                   int64
discount_pct           float64
returns                  int64
revenue                float64
dtype: object

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10020 entries, 0 to 10019
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   customer_id          10020 non-null  str    
 1   region               10020 non-null  str    
 2   customer_type        10020 non-null  str    
 3   acquisition_channel  10020 non-null  str    
 4   tenure_months        10020 non-null  int64  
 5   orders               10020 non-null  int64  
 6   discount_pct         9970 non-null   float64
 7   returns              10020 non-null  int64  
 8   revenue              10020 non-null  float64
dtypes: float64(2), int64(3), str(4)
memory usage: 704.7 KB


In [6]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
customer_id,10020,10000,C000440,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
region,10020,4,South,3820,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_type,10020,3,Consumer,6443,NaN,NaN,NaN,NaN,NaN,NaN,NaN
acquisition_channel,10020,5,Organic,2963,NaN,NaN,NaN,NaN,NaN,NaN,NaN
tenure_months,10020.0,NaN,NaN,NaN,21.788822,14.492937,1.0,11.0,19.0,29.0,72.0
orders,10020.0,NaN,NaN,NaN,6.5502,4.618187,1.0,3.0,6.0,9.0,39.0
discount_pct,9970.0,NaN,NaN,NaN,8.889589,4.691856,0.0,5.4,8.5,12.3,24.6
returns,10020.0,NaN,NaN,NaN,0.259381,0.556685,0.0,0.0,0.0,0.0,7.0
revenue,10020.0,NaN,NaN,NaN,20093.946462,33116.172833,750.84,4889.3725,9664.81,18848.6675,415741.14


In [7]:
missing_summary = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_pct": df.isna().mean() * 100
})

missing_summary.sort_values(
    "missing_count",
    ascending=False
)

,missing_count,missing_pct
discount_pct,50,0.499002
customer_id,0,0.000000
region,0,0.000000
acquisition_channel,0,0.000000
customer_type,0,0.000000
tenure_months,0,0.000000
orders,0,0.000000
returns,0,0.000000
revenue,0,0.000000


In [8]:
df.duplicated().sum()

np.int64(20)

In [9]:
df["customer_id"].duplicated().sum()

np.int64(20)

In [10]:
duplicate_customers = df[
    df["customer_id"].duplicated(keep=False)
].sort_values("customer_id")

duplicate_customers

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue
439,C000440,East,Consumer,Organic,13,1,6.1,0,1343.35
10009,C000440,East,Consumer,Organic,13,1,6.1,0,1343.35
576,C000577,West,SMB,Paid Search,30,10,14.8,1,27712.24
10006,C000577,West,SMB,Paid Search,30,10,14.8,1,27712.24
582,C000583,North,Consumer,Organic,25,5,6.3,0,8137.00
10016,C000583,North,Consumer,Organic,25,5,6.3,0,8137.00
1731,C001732,North,Consumer,Social,55,14,12.3,0,16644.34
10002,C001732,North,Consumer,Social,55,14,12.3,0,16644.34
2750,C002751,South,Enterprise,Social,11,8,12.2,0,66700.47
10010,C002751,South,Enterprise,Social,11,8,12.2,0,66700.47


In [11]:
invalid_orders = df[df["orders"] < 0]

invalid_orders

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue


In [12]:
invalid_returns = df[
    (df["returns"] < 0) |
    (df["returns"] > df["orders"])
]

invalid_returns

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue


In [13]:
invalid_discount = df[
    (df["discount_pct"] < 0) |
    (df["discount_pct"] > 100)
]

invalid_discount

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue


In [14]:
invalid_revenue = df[df["revenue"] < 0]

invalid_revenue

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue


In [15]:
invalid_tenure = df[df["tenure_months"] <= 0]

invalid_tenure

,customer_id,region,customer_type,acquisition_channel,tenure_months,orders,discount_pct,returns,revenue


In [16]:
df["revenue_per_order"] = (
    df["revenue"] / df["orders"]
)

In [17]:
df["revenue_per_order"].describe()

count    10020.000000
mean      2490.640008
std       2254.125064
min        574.415000
25%       1313.448611
50%       1643.461667
75%       2676.824405
max      57007.400000
Name: revenue_per_order, dtype: float64

In [18]:
df.nlargest(10, "revenue_per_order")[
    [
        "customer_id",
        "customer_type",
        "region",
        "orders",
        "revenue",
        "revenue_per_order",
        "discount_pct",
        "tenure_months",
        "returns"
    ]
]

,customer_id,customer_type,region,orders,revenue,revenue_per_order,discount_pct,tenure_months,returns
7794,C007795,SMB,East,7,399051.80,57007.400000,7.2,19,0
7456,C007457,SMB,East,3,120381.20,40127.066667,12.5,9,0
8511,C008512,Consumer,South,6,192199.20,32033.200000,8.8,29,0
8321,C008322,Consumer,South,9,242573.80,26952.644444,1.8,25,1
6611,C006612,Consumer,South,4,95118.60,23779.650000,18.4,26,1
9377,C009378,Enterprise,North,10,159676.14,15967.614000,5.9,24,2
6821,C006822,Enterprise,West,11,162311.59,14755.599091,7.6,32,1
251,C000252,Enterprise,West,10,146553.69,14655.369000,2.3,20,0
6420,C006421,Enterprise,South,11,155553.40,14141.218182,6.1,13,1
8509,C008510,Enterprise,West,13,183776.37,14136.643846,8.9,19,1


## Data Quality Assessment

### Dataset Profile

- **Rows:** 10,020
- **Columns:** 9
- **Grain:** One row is intended to represent one customer.
- **Business Key:** `customer_id`
- **Unique customers:** 10,000

The dataset contains 10,020 records but only 10,000 unique customer IDs, indicating that duplicate customer records are present.

The dataset contains:
- 4 categorical attributes: `customer_id`, `region`, `customer_type`, `acquisition_channel`
- 5 numerical attributes: `tenure_months`, `orders`, `discount_pct`, `returns`, `revenue`

### Missing Values

- `discount_pct` contains **50 missing values**, representing approximately **0.50%** of the dataset.
- All other columns contain no missing values.
- The missing discount values should not be blindly imputed before understanding their impact on discount-based analysis.

### Duplicate Records

- **20 completely duplicated rows** were identified.
- **20 duplicate `customer_id` records** were identified.
- Since the intended grain is one row per customer, these duplicates violate the expected business grain.
- The duplicated records appear to be exact copies of existing customer records rather than separate customer activity.

Therefore, the duplicate records should be investigated and handled before calculating customer-level aggregates or business metrics.

### Business Rule Violations

- **Orders:** No invalid negative order values were identified.
- **Returns:** No negative returns or cases where returns exceeded orders were identified.
- **Discount:** No values outside the expected 0–100% range were identified.
- **Revenue:** No negative revenue values were identified.
- **Tenure:** No customers with zero or negative tenure were identified.

The core numerical fields therefore satisfy the basic business validity rules tested.

### Business Sanity Checks

A derived metric, `revenue_per_order`, was calculated to investigate whether revenue is reasonable relative to customer order volume.

The distribution showed:

- Mean revenue per order: approximately **₹2,490.64**
- Median revenue per order: approximately **₹1,643.46**
- Maximum revenue per order: approximately **₹57,007.40**

The large difference between the mean and median, together with the very high maximum, indicates a highly right-skewed distribution and the presence of potential high-value or anomalous observations.

The highest revenue-per-order customer generated approximately ₹57,007 per order, which is substantially higher than the typical customer. This should be investigated further during the outlier-analysis phase rather than immediately classified as bad data.

### Overall Assessment

The dataset is **usable for further analysis, but should not yet be treated as fully clean**.

The primary data-quality issues identified are:

1. **20 duplicate customer records**
2. **50 missing `discount_pct` values**
3. **Potential extreme revenue-per-order observations**

The basic business-rule validation did not identify invalid orders, returns, discounts, revenue, or tenure values.

The duplicate customer records should be handled before customer-level analysis because they can artificially inflate customer counts and revenue totals. Missing discount values should be retained or handled according to the analytical question rather than arbitrarily imputed.

The extreme revenue-per-order observations should be investigated separately to determine whether they represent legitimate high-value customers or potential data-quality issues.

**Conclusion:** The dataset is sufficiently reliable to continue the investigation after documenting the duplicate-record issue, while the missing values and potential outliers should be explicitly accounted for in subsequent analysis.